itests/fevm_test.go

```diff
@@ -1436,3 +1436,78 @@ func TestEthGetBlockByNumber(t *testing.T) {
 	require.NotNil(t, pendingBlock)
 	require.True(t, pendingBlock.Number >= latest)
 }
+
+func TestEthGetTransactionByBlockHashAndIndex(t *testing.T) {
+	ctx, cancel, client := kit.SetupFEVMTest(t)
+	defer cancel()
+
+	key, ethAddr, filAddr := client.EVM().NewAccount()
+	// Send some funds to the f410 address
+	kit.SendFunds(ctx, t, client, filAddr, types.FromFil(10))
+
+	// Deploy the Blocktest contract
+	tx := deployContractWithEth(ctx, t, client, ethAddr, "./contracts/MultipleEvents.hex")
+	client.EVM().SignTransaction(tx, key.PrivateKey)
+	hash := client.EVM().SubmitTransaction(ctx, tx)
+
+	receipt, err := client.EVM().WaitTransaction(ctx, hash)
+	require.NoError(t, err)
+	require.NotNil(t, receipt)
+	require.Equal(t, ethtypes.EthUint64(1), receipt.Status)
+
+	// Use the block hash from the receipt
+	blockHash := receipt.BlockHash
+
+	// Get the block by its hash
+	block, err := client.EthGetBlockByHash(ctx, blockHash, false)
+	require.NoError(t, err)
+	require.NotNil(t, block)
+
+	ethTx, err := client.EthGetTransactionByBlockHashAndIndex(ctx, block.Hash, ethtypes.EthUint64(0))
```

Comment:
This is potentialy flaky because it's not certain that you're going to find the transaction you want at position `0` here - it could be that one a miner that's submitting windowposts to keep the chain alive submits something at this same block, so you can't be sure unfortunately.

I think the better approach for these tests would be similar to `TestMultipleEvents` in eth_filter_test.go which sets up a scenario where it's sure that all transactions got into the same tipset before it proceeds to check them. So in here, you'd do submit your transaction (ideally multiple so you can test more than `0`), confirm that the tipset is shaped the way you want, and then proceed to verify. You could inspect a tipset to see the messages included in the way you want, then once you've got what you need then do your queries with your new API and continue.

It would also be OK if you did both of these APIs in a single test case too, it'll be cheaper to run that way too!

Reply:
You should also test the case where there's a tipset but no messages, what should that return? And where the index is >= number of messages. Let's ensure we have the error cases covered and get roughly the message we expect (you can use `ErrorContains` to check just a portion of the message you care about).